src/Template/PhpTemplate.php

```diff
@@ -0,0 +1,232 @@
+     * It MAY throw {@see Exception} if there is a problem retrieving the value.
+     *
+     * @param array|ArrayAccess|ContainerInterface $context The context to get the retrieval function for.
+     *
+     * @throws Exception If problem retrieving.
+     *
+     * @return callable The function that retrieves a context value.
+     * function (string $key, mixed $default): mixed
+     */
+    protected function getContextFunction($context): callable
+    {
+        return function (string $key, $default = null) use ($context) {
+            return $this->getContextValue($context, $key, $default);
+        };
+    }
+
+    /**
+     * Retrieves a function which itself invokes a template function by name.
+     *
+     * That function MUST take a string key, and a variable list of arguments.
+     * It MUST invoke an underlying function mapped by name.
+     * It MUST pass these arguments to the underlying function , as a variable list.
+     * It MUST return the result of the invocation of the underlying function.
+     * It MAY throw an {@see Exception} if there is a problem invoking the underlying function.
+     *
+     * @param array|ArrayAccess|ContainerInterface $context The context to get the functions function for.
+     *
+     * @throws Exception If problem retrieving.
+     *
+     * @return callable The function that invokes a template function.
+     * function (string $funcName, ...$args): mixed
+     */
+    protected function getFunctionsFunction($context): callable
+    {
+        return function (string $funcName, ...$args) use ($context) {
+            return $this->invokeFunction($funcName, $args);
+        };
+    }
+
+    /**
+     * Retrieves a value from a context.
+     *
+     * @param array|ArrayAccess|ContainerInterface $context The context to retrieve the value from.
+     * @param string|Stringable $key The key to retrieve the value for.
+     * @param mixed $default The value to retrieve if the key is not found.
+     * @throws Exception If problem retrieving value.
+     *
+     * @return mixed The value that corresponds to the specified key in the context, or the default value if not found.
+     */
+    protected function getContextValue($context, string $key, $default = null)
+    {
+        try {
+            $this->validateContext($context);
+        } catch (InvalidArgumentException $e) {
+            throw new InvalidArgumentException($this->__('Could not retrieve value for key "%1$s": invalid context', $key), 0, $e);
+        }
+
+        if ($context instanceof ContainerInterface) {
+            if (!$context->has($key)) {
+               if (!array_key_exists($key, $this->defaultContext)) {
+                   return $default;
+               }
+
+               return $this->defaultContext[$key];
+            }
+
+            return $context->get($key);
+        }
+
+        // It's an array-like
+        {
+            if (!array_key_exists($key, $context)) {
```

Comment:
Be advised that `array_key_exists` on objects, even if they implement `ArrayAccess`, is [deprecated as of PHP 7.4](https://3v4l.org/iRHCK).

Reply:
That is [exactly right][array_key_exists_with_objects]. Looks like I have to add explicit support.

[array_key_exists_with_objects]: https://wiki.php.net/rfc/deprecations_php_7_4#array_key_exists_with_objects

Reply:
I added tests for this via data provider, but I am unable to reproduce this issue due to the fact that PHPUnit 6 does not support PHP 7.4, and gives an error before the above code can run.

Reply:
Here's what's strange, though: If `array_key_exists()` has no support for `ArrayAccess`, and "only works on mangled properties", then why does this code work even on PHP 7.3 and below?

Reply:
[Because public properties are not name-mangled](https://3v4l.org/jqO9a), and in order for `$context` to be useful in this context (pun definitely intended) the properties need to be public.

Unless they are magic, in which case `array_key_exists` will definitely not work.

Reply:
How does the fact that `array_key_exists()` seems to work on `ArrayAccess` relate to public properties and their names?

Reply:
It doesn't. I replied to your question.

> If array_key_exists() has no support for ArrayAccess, and "only works on mangled properties", then why does this code work even on PHP 7.3 and below?

The code works on PHP 7.3 and below because:

* It [does work](https://3v4l.org/ufn9L) for `ArrayAccess`
* It [does work](https://3v4l.org/cA5XI) for `object` (as long as the properties are public, because their mangled names would be the same as the original names)

Reply:
I get that it works, but this seems to contradict the documentation.

Reply:
Anyhow, fixed.